### DISCUSSION TITLE
Convergence of advection in PF module

### DISCUSSION THREAD

**Discussion by lloyd-5731** [idaholab/moose#23082] [Q&A Modules: General] [answered]
I am looking to advect a variable inside the porous flow module.
The value has to be limited between 0 and 1, however I am experiencing massive overshooting and instability (resulting in oscillations) when advecting, leading it to be outside this region.
I have looked at the different stabiliser options, however none seem to be doing the trick. The closest to realistic results I am getting are from the `PorousFlowBasicAdvection` kernel. It seems logical the upwinding will not work because there are regions where the value is approximately in a steady state, I do not understand why the KT stabilisation is not helping however.
In the wider context, this variable will be acted on by other kernels but these have been removed for the MWE below.

Some of the advection stabilisation options I've tried have been commented out in the MWE.
Any direction would be appreciated.

Best wishes,
Lloyd

MWE:
```
P0 = 5e5      # Injection pressure
rho = 920       # Density
k = 2.5e-11       # Permeability
phi = 0.45      # Porosity
T = 298

init_variable = 0  
input_from_BC = 1


[Mesh]
    [plate]
        type = GeneratedMeshGenerator
        dim = 2
        xmax = 0.01
        nx = 5
        ymax = 1.0
        ny = 100
    []
[]
[UserObjects]
    [dictator]
        type = PorousFlowDictator
        porous_flow_vars = 'porepressure'
        number_fluid_phases = 1
        number_fluid_components = 1
    []
    [pc]
        type = CustomCapillaryPressure
    []
    [fluid_advective_flux]
        type = PorousFlowAdvectiveFluxCalculatorUnsaturated
        flux_limiter_type = superbee
    []
[]

[GlobalParams]
    PorousFlowDictator = dictator
    gravity = '0 0 0'
[]

[Variables]
    [porepressure]
        initial_condition = 0
    []
    [variable_to_advect]
        initial_condition = ${init_variable}
    []
[]

[Kernels]
    [dmdt]
        type = PorousFlowMassTimeDerivative
        variable = porepressure
    []
    [flux]
        type = PorousFlowAdvectiveFlux
        variable = porepressure
    []
    [advect_var_flux]
        type = PorousFlowBasicAdvection
        # type = PorousFlowAdvectiveFlux
        # type = FluxLimitedTVDAdvection
        # type = PorousFlowFluxLimitedTVDAdvection
        
        variable = variable_to_advect
        
        # advective_flux_calculator = fluid_advective_flux
    []
    [advect_var_time_derivative]
        type = ADTimeDerivative
        # type = MassLumpedTimeDerivative
        
        # multiply_by_density = false
        
        variable = variable_to_advect
    []
[]

[AuxVariables]
    [saturation]
        family = MONOMIAL
        order = CONSTANT
    []
[]

[AuxKernels]
    [saturation]
        type = PorousFlowPropertyAux
        variable = saturation
        property = saturation
    []
[]

[BCs]
    [injection]
        type = DirichletBC
        variable = porepressure
        boundary = 'top'
        value = ${P0}
    []

    [DOC_in]
        type = DirichletBC
        variable = variable_to_advect
        boundary = 'top'
        value = ${input_from_BC}
    []
[]

[Modules]
    [FluidProperties]
        [resin]
            type = SimpleFluidProperties
            density0 = ${rho}
        []
    []
[]

[Materials]
    [porosity]
        type = PorousFlowPorosity
        porosity_zero = ${phi}
    []
    [permeability]
        type = PorousFlowPermeabilityConst
        permeability = '${k} 0 0   0 ${k} 0   0 0 ${k}'
    []
    [saturation_calculator]
        type = PorousFlow1PhaseP
        porepressure = porepressure
        capillary_pressure = pc
    []
    [temperature]
        type = PorousFlowTemperature
        temperature = ${T}
    []
    [massfrac]
        type = PorousFlowMassFraction
    []
    [resin]
        type = PorousFlowSingleComponentFluid
        fp = resin
        phase = 0
    []
    [relperm]
        type = PorousFlowRelativePermeabilityCorey
        n = 3
        s_res = 0.1
        sum_s_res = 0.1
        phase = 0
    []
    [darcy_velocity]
        type = PorousFlowDarcyVelocityMaterial
    []
[]

[Preconditioning]
    [smp]
        type = SMP
        full = true
        petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
        petsc_options_value = ' lu       mumps'
    []
    
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    end_time = 2000

    [TimeStepper]
        type = SolutionTimeAdaptiveDT
        dt = 0.001
    []

    automatic_scaling = true
    nl_abs_tol = 1e-9
    nl_rel_tol = 1e-9
[]

[Outputs]
    exodus = true
    print_linear_residuals = false
[]
```

**Comment by lloyd-5731** [comment]
Here is an example of it working well with the `PorousFlowBasicAdvection` and `ADTimeDerivative` :
<img width="659" alt="Screenshot 2023-01-05 at 16 33 46" src="https://user-images.githubusercontent.com/116571080/210831937-0a9295ed-85b0-4bc7-a797-fcf58eccd7a8.png">

And here is an example of the advection acting strangely with `PorousFlowFluxLimitedTVDAdvection` and `MassLumpedTimeDerivative` using `superbee` which to my understanding should be better:
<img width="898" alt="Screenshot 2023-01-05 at 16 36 37" src="https://user-images.githubusercontent.com/116571080/210832536-56a26743-7cd9-4a89-a976-b19968922a04.png">

in a post from @hugary1995 #21598 he mentions `-pc_type lu` should not be used with `DirichletBC`. Could this be a source of the error.

**Comment by WilkAndy** [comment, marked as answer]
The PorousFlow TVD schemes are mostly designed to work within a PorousFlow context only, and i can't really guarantee that they'll work outside of it.   Your current setup has PorousFlow creating the advective velocity, but a **separate** variable advecting via that velocity.   If that separate variable is supposed to be a fluid component, then i'd make it such (using `number_fluid_components = 2` in the `Dictator` and following something like modules/porous_flow/test/tests/flux_limited_TVD_pflow/pffltvd_1D.i , but probably using the https://mooseframework.inl.gov/source/actions/PorousFlowUnsaturated.html Action instead of Kernels).  However, if the separate variable is not a fluid component, then I think i'd use the https://mooseframework.inl.gov/source/kernels/ConservativeAdvection.html Kernel

**Reply by lloyd-5731** [reply]
Hi Andy, 
Thanks for the comment and the help!

1) The variable is an attribute of the fluid that im looking to advect at the Darcy velocity, rather than a separate fluid component so it is my understanding that using a different fluid component wouldn't work. Is this right?

2) Unfortunately I don't think I can use the action system as I need to add my custom kernels in, as this is only a MWE>

3) I will have a look at `ConservativeAdvection`, will this still advect at the Darcy velocity as it does when I use the porous flow advection kernels?

Ill get back to you and let you know the outcome!

Lloyd

**Reply by WilkAndy** [reply]
(1) You could treat it as a fluid component, and have all the fluid properties, etc, not depend on it, and PorousFlow will work fine, BUT it seems an overkill - i think your current approach is better

(3) Yes, `ConservativeAdvection` will advect at the Darcy velocity - you just have to define an AuxVariable that is the Darcy velocity.   The slight problem is that you'll miss terms in the Jacobian, because `ConservativeAdvection` won't calculate the derivatives of the Darcy velocity with respect to the porepressure, so you'll get poorer convergence.  However, in many real-life situations, the porepressure rapidly goes to steadystate, so the Darcy velocity doesn't change much anyway, so it's likely you won't see many problems, except for during the first few timesteps.

A

**Reply by lloyd-5731** [reply]
Thanks Andy, that's made a real difference and is looking much better.
I have slightly changed the `ConservativeAdvection` kernel so that it takes the Darcy velocity as the velocity (scaled by 1/ porosity) and it is now advecting with the flow front.

The instability at the start however is causing the instability to ripple through the rest of the solve.

I suppose the solution to this is to write the proper Jacobian, but im not sure why the existing Jacobian is incorrect is I have just replace the viscosity with the Darcy Velocity?

The relevant part of the .C file looks like this:

```
 Real
LSConservativeAdvection::negSpeedQp() const
{
  return -_grad_test[_i][_qp] * _darcy_velocity[_qp][0] / _porosity;
}

Real
LSConservativeAdvection::computeQpResidual()
{
  // This is the no-upwinded version
  // It gets called via Kernel::computeResidual()
  return negSpeedQp() * _u[_qp];
}

Real
LSConservativeAdvection::computeQpJacobian()
{
  // This is the no-upwinded version
  // It gets called via Kernel::computeJacobian()
  return negSpeedQp() * _phi[_j][_qp];
}
```

Thanks for the help,
Lloyd

**Reply by WilkAndy** [reply]
I'm glad that works.  Regarding your `LSConservativeAdvection`, I think that it would be better to do one of the following

1. Inherit from `ConservativeAdvection`, and just override the single function `negSpeedQp` by including your porosity.  No other coding should be necessary.  Then you get all the goodness of `ConservativeAdvection`, viz the tests that guarantee it is correct, and the upwinding.   (Perhaps this is what you've done already - i couldn't really tell from the above code snippets.)
2. Use `ConservativeAdvection` (ie, don't do any C++ coding at all) but create an AuxVariable that is Darcy / porosity.   I think that's what i'd do, as it keep code modifications to a minimum (actually, zero modifications in this case!).

I think the instability is due to not using `upwinding_type = full`, but i could easily be wrong!

**Comment by lloyd-5731** [comment]
Hi Andy, I had previously had 1) implemented initially but I have now implemented 2) (`upwinding_type = full` is activated)

I have realised that the source of my error is that because I am saturating the fibre, the darcy velocity is zero in the unsaturated part, the advection velocity to be discontinuous.

I am trying to come up a plan to tackle this. My current idea is to create a subapp that slightly diffuses the darcy velocity at each time step to make it continuous. - Is this reasonable or would you recommend something else?

When trying to do this I am currently taking out the components of the darcy velocity, passing them into my sub app and then trying to diffuse them. This won't work however, I think because the `PorousFlowDarcyVelocityComponent` is a `Constant Monomial`, and therefore I cannot get the gradient for the diffusion.

I'd love to hear your thoughts.

Lloyd

**Reply by WilkAndy** [reply]
What do you think happens in real life, @lloyd-5731 ?

**Reply by lloyd-5731** [reply]
So I think that the variable should advect with the saturation front as it is a property of the fluid. Obviously it doesn't diffuse in reality, as it is a property of the fluid and the fluid is not diffusing, but it seems to be a work around at the minute. Now I have that working Im going to explore a couple of different more representative methods 👍
